runtime/src/lib.rs

```diff
@@ -198,22 +219,47 @@ pub unsafe extern "C" fn cairo_native__dict_drop(
 /// definitely unsafe to use manually.
 #[no_mangle]
 pub unsafe extern "C" fn cairo_native__dict_dup(
-    ptr: *mut FeltDict,
-    dup_fn: extern "C" fn(*mut c_void) -> *mut c_void,
+    old_dict: &FeltDict,
+    dup_fn: Option<extern "C" fn(*mut c_void, *mut c_void)>,
 ) -> *mut FeltDict {
-    let old_dict = &*ptr;
     let mut new_dict = Box::new(FeltDict {
-        inner: HashMap::default(),
+        mappings: HashMap::with_capacity(old_dict.mappings.len()),
+
+        layout: old_dict.layout,
+        elements: if old_dict.mappings.is_empty() {
+            null_mut()
+        } else {
+            alloc(Layout::from_size_align_unchecked(
+                old_dict.layout.pad_to_align().size() * old_dict.mappings.len(),
+                old_dict.layout.align(),
+            ))
+            .cast()
+        },
+
+        // TODO: Check if `0` is fine or otherwise we should copy the value from `old_dict` too.
         count: 0,
-        free_fn: old_dict.free_fn,
     });
 
-    new_dict.inner.extend(
-        old_dict
-            .inner
-            .iter()
-            .filter_map(|(&k, &v)| (!v.is_null()).then_some((k, dup_fn(v)))),
-    );
+    for (&key, &old_index) in old_dict.mappings.iter() {
+        let old_value_ptr = old_dict
+            .elements
+            .byte_add(old_dict.layout.pad_to_align().size() * old_index);
+
+        let new_index = new_dict.mappings.len();
```

Comment:
Why not just using the `enumerated()` method to get the current index?

Reply:
I missed it. Fixed.